### PR TITLE
Improve compilation

### PR DESCRIPTION
**Pull request by mememori**
Note about last commit:

`HELF.Event` will always depend on compile-time on the event struct module and on the event handler. But the way things are organized, if an event is recompiled, only the dispatcher has to be recompiled. Eg:
```
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Software.Service.Event.Decryptor.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Event.Dispatcher.beam
```
So i am okay with it

About PK, since it defined lot of module atoms, elixir made it depend on the models (and in exchange all models that used autogenerated pks had compile-time dependency on pk), so, whenever you changed a model that used autogenerated pks, you would recompile that model, the controllers/services/etc that depended on that model, the pk module, all models that used autogenerated pks and every controllers/services/etc that depended on those models. By not having implicit potential dependency on the models (by using generic atoms), this issue was solved and now when you modify a model that uses pk, you will probably only have to recompile a few modules (instead of ~50).

I also added an util to generate dependency graphs. It's very useful to see the overview of system dependency, the event-related dependency graph etc. It also generates a .dot file and store it. I did it this way because i hope that, in the future, our jenkins pipeline could compare this generated dot file with master's generated dot file, that way we can give a second look before accepting reviews as it simply overviews what was introduced through a dependency graph (i believe it could help us to avoid potential bugs)

---
This change is [Reviewable](https://reviewable.io/reviews/hackerexperience/helix/124)

### CODE REVIEW

**Comment by mememori** [comment]
Also, as could be noticed, this wasn't a planned task; i had to do it when i started writing network models and had to recompile ~50 modules every time :P but i think it was worth the time

File: lib/process/model/process/process_type.ex [hunk 1]
```diff
@@ -7,7 +7,7 @@ defprotocol Helix.Process.Model.Process.ProcessType do
   @spec dynamic_resources(t) :: [resource]
   def dynamic_resources(data)
 
-  @spec event(t, Process.t, :created | :completed) :: [Helix.Event.t]
+  @spec event(t, Process.t, :created | :completed) :: [struct]
```
Comment (mememori) [inline lib/process/model/process/process_type.ex]: About this part, it's funny:

- Runtime dependencies don't cause your code to need to be recompiled if the dependency is modified
- Compiletime dependencies do
- If you have a compiletime dependency that has a runtime-dependency, you indirectly has that dependency but since it could affect the code generated on compiletime, it works just like a compiletime-dependency and thus you have to recompile your module whenever that dependency (that was not compiletime-dep of anything) changes
- Protocols are based around macros, so they are compiletime dependencies of their implementations
- Since adding a type implies into runtime dependency and protocols are compiletime dependencies, this implies that the external referenced type now works like a compiletime dependency, thus, any time that any event handler was modified, so were all ProcessType implementations and, by consequence, any module that was depending on them on certain conditions.

tl;dr, before this change, any event handler change would cause:
```
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Controller.Process.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Controller.TableOfProcesses.Allocator.Plan.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Event.Dispatcher.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Software.Encryptor.ProcessType.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Atom.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Software.Decryptor.ProcessType.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Software.Encryptor.ProcessType.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Software.Decryptor.ProcessType.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.BitString.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Float.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Function.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Integer.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.List.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Map.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.PID.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Port.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Reference.beam
_build/dev/lib/helix/ebin/ MODIFY Elixir.Helix.Process.Model.Process.ProcessType.Tuple.beam
```

Comment (mememori) [inline lib/process/model/process/process_type.ex]: Also, naturally, this problem won't happen as soon as we add that type on HELF as helf won't be recompiled on every event handler added :stuck_out_tongue:
